wai-extra/Network/Wai/Middleware/Rewrite.hs

```diff
@@ -1,32 +1,305 @@
+-- paths formulation.
+--
+-- > {-# LANGUAGE ViewPatterns #-}
+-- >
+-- > rewritePathFromPhp :: Middleware
+-- > rewritePathFromPhp = rewritePureWithQueries pathFromPhp
+-- >
+-- > pathFromPhp :: PathsAndQueries -> H.RequestHeaders -> PathsAndQueries
+-- > pathFromPhp (pieces, queries) _ = piecesConvert pieces queries
+-- >     where
+-- >         piecesConvert :: [Text] -> H.Query -> PathsAndQueries
+-- >         piecesConvert ["index.php"] qs@(join . lookup "page" -> Just page) =
+-- >             ( ["index", TE.decodeUtf8With TE.lenientDecode page]
+-- >             , delete ("page", pure page) qs
+-- >             )
+-- >         piecesConvert ps qs = (ps, qs)
+--
+-- On the other side, will use 'rewriteRequestPure' to rewrite outgoing
+-- requests to the original website from the reverse proxy code (using the
+-- 'Network.HTTP.ReverseProxy.WPRModifiedRequest' or
+-- 'Network.HTTP.ReverseProxy.WPRModifiedRequestSecure' constructors. Note,
+-- these links will only work if the haddock documentation for
+-- <https://hackage.haskell.org/package/http-reverse-proxy http-reverse-proxy>
+-- is installed).
+--
+-- > rewritePhpFromPath :: Request -> Request
+-- > rewritePhpFromPath = rewriteRequestPure phpFromPath
+-- >
+-- > phpFromPath :: PathsAndQueries -> H.RequestHeaders -> PathsAndQueries
+-- > phpFromPath (pieces, queries) _ = piecesConvert pieces queries
+-- >     where
+-- >         piecesConvert :: [Text] -> H.Query -> PathsAndQueries
+-- >         piecesConvert ["index", page] qs = ( ["index.php"], ("page", pure . TE.encodeUtf8 $ page) : qs )
+-- >         piecesConvert ps qs = (ps, qs)
+--
+-- For the whole example, see
+-- <https://gist.github.com/dbaynard/c844d0df124f68ec8b6da152c581ce6d>.
 
--- staticConvert :: [Text] -> H.RequestHeaders -> [Text]
--- staticConvert pieces _ = piecesConvert pieces
+-- $pathsandqueries
+-- /e.g./ /\/foo\/bar/ would look like
+--
+-- > ["foo", "bar"] :: Text
+--
+-- /?bar=baz/ would look like
+--
+-- > [("bar", Just "baz")] :: QueryText
+--
+-- Together,
+--
+-- /\/foo?bar=baz/ would look like
+--
+-- > (["foo"],[("bar", Just "baz")]) :: PathsAndQueries
+
+-- $upgrading
+-- It is quite simple to upgrade from 'rewrite' and 'rewritePure', to
+-- 'rewriteWithQueries' and 'rewritePureWithQueries'.
+-- Insert 'Data.Bifunctor.first' as the following example demonstrates.
+--
+-- Old versions of the libary could only handle path pieces, not queries.
+-- This could have been supplied to 'rewritePure'.
+--
+-- @
+-- staticConvert' :: [Text] -> H.RequestHeaders -> [Text]
+-- staticConvert' pieces _ = piecesConvert pieces
+--   where
+--    piecesConvert [] = ["static", "html", "pages.html"]
+--    piecesConvert route@("pages":_) = "static":"html":route
+-- @
+--
+-- Instead, use this function, supplied to 'rewritePureWithQueries'.
+--
+-- @
+-- staticConvert :: PathsAndQueries -> H.RequestHeaders -> PathsAndQueries
+-- staticConvert pathsAndQueries _ = first (piecesConvert pathsAndQueries)
```

Comment:
Yes. Also there’s a typo 👍 .

To clarify, previous users of the library will have code that looks like `staticConvert'`; this was the example given in the old formulation of the library. That code will still work. However, to migrate to using the new functions they just need to add the single function, `first :: ([Text] -> [Text]) -> PathsAndQueries -> PathsAndQueries` (I’ll add the specialised type to the docs). `pathsOnly` has to handle the `RequestHeaders` parameter too; first does not do this.